### PR TITLE
Roll AFL++

### PR DESCRIPTION
**Pull request by jonathanmetzman**

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Review by oliverchang — APPROVED** [review]
awesome to see the CI in action!

**Comment by oliverchang** [comment]
Is there a way to automatically test more projects at once?

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
One weird thing I noticed is that the base-image builds are happening in us-central1 in trial builds instead of in global like the production builds.

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py skcms --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine afl

**Comment by jonathanmetzman** [comment]
The AFL++ roll breaks the following:
[woff2](https://00f74ba44bd13595e26e7c77dd54e516c6b4a6749b-apidata.googleusercontent.com/download/storage/v1/b/oss-fuzz-gcb-logs/o/log-c400affc-9e20-4620-8143-0127854bedd4.txt?jk=AFshE3Xhy4n6_1QkhfBvMQNdeQVN7epb2G22v5qtKPT3WrQPtywLGnLLvjtOG3fT5Nx7lViMFDktjLS6iasdpAnczeuE9wUlJ_NRXLnfl74ZnBrWQbTf3L9ukJllrfppg-WFfwRneGAkAXeGMTWp9TiDaVjvId224M6Fo2dctNBX94KtDupsPy_pdM5I3fxMHKWKdrbu4AezI3XVO-XWL20qdpzy6WMQHTQzvPFD-c13e1wirCMjybfvorJWlHK912BX0Gij7zN2erNlSN8utfq8Iv9edF_u_d769zEBIo3xCHag0SsicEDbxF5T0kgKuRQfmguMlGurRAUiVoVdxS5xKQkcPRwyaWUtD8WeIiArAa6g_Hu7DJOcV_yJtWJkC4Jm4Q_E-saNBaCyhpwkBXZ7WrB_1QZTiMPpp3qg_4a_CTyDxB7_xZOkhXKBOic3GHQIEZyl8gLteiPThqmShUcJGA5skAL3BhWhTR0PxhAZu2L8NhYGjoFiA3CV9E6fshPt6oivijrfxjUTNTLXiqTylI7I-aZmUzvm3_z6BLKpvNaYj-u0vDhDOl3WreeAgHwM3amBWHcDOR7eAUm8kw7RhrNbS25hjT2rANbllv6TNrAHf8NCiGUUFJiMdBH7VZVw4GqGwxP41hzio_Z8ORavGzOb6m904KkDJdDfOgtRZcuZ9LRwW9_TES5YdhotEJzZjivkpCuOhOMz9F6rGLy2KMoa9uJQPbj4j5FCB28ImDHpuXGJYAytBtcHkhvdedVbhARpViVRRWnCFeCFstvlIRwuYoFk599HAt1qUXHq6trBDerydli82wJp5ekVDgC_d_cOBD4FQrujvy0J_hslqz2DI7ViyvEF7-5spwvaRV2igAguw9_aTWSwt8jeV4IOWFenmHv7OXt8XvjFCRbOSABwQpY71Y7dIecBr3yUun9t0B3jvHD9TMreEjjKSYI2c_it7UyXU1H72qsWMYgPYAJWwvBG5xm_VDexYD4UtNWjVNSeTnTBhGxLKbbFBT4C4rnOjazwynOcVGVc8li1&isca=1)

**Comment by jonathanmetzman** [comment]
/gcbrun trial_build.py all --sanitizer address --fuzzing-engine afl

File: infra/build/functions/build_project.py [hunk 1]
```diff
@@ -97,6 +97,20 @@ def get_project_data(project_name):
...
 
+def get_sanitizer_strings(sanitizers):
+  """Accepts the sanitizers field from project.yaml where some sanitizers can be
+  defined as experimental. Returns a list of sanitizers."""
+  processed_sanitizers = []
+  for sanitizer in sanitizers:
+    if isinstance(sanitizer, six.string_types):
+      processed_sanitizers.append(sanitizer)
+    elif isinstance(sanitizer, dict):
+      for key in sanitizer.keys():
```
Comment (jvoisin) [inline infra/build/functions/build_project.py:108]: `processed_sanitizers.extend(sanitizer.keys())` is better :)

**Review by jvoisin — APPROVED** [review]

**Comment by jonathanmetzman** [comment]
Only woff2 build fails:
```
Step #15 - "compile-afl-address-x86_64":         -c c/enc/block_splitter.c -o bin/obj/c/enc/block_splitter.o
Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
Step #15 - "compile-afl-address-x86_64":         -c c/enc/memory.c -o bin/obj/c/enc/memory.o
Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
Step #15 - "compile-afl-address-x86_64":         -c c/enc/compress_fragment_two_pass.c -o bin/obj/c/enc/compress_fragment_two_pass.o
Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
Step #15 - "compile-afl-address-x86_64":         -c c/enc/backward_references_hq.c -o bin/obj/c/enc/backward_references_hq.o
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
Step #15 - "compile-afl-address-x86_64":         -c c/enc/entropy_encode.c -o bin/obj/c/enc/entropy_encode.o
Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
Step #15 - "compile-afl-address-x86_64":         -c c/enc/bit_cost.c -o bin/obj/c/enc/bit_cost.o
Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
Step #15 - "compile-afl-address-x86_64": rm -f libbrotli.a
Step #15 - "compile-afl-address-x86_64": ar -crs libbrotli.a bin/obj/c/common/transform.o bin/obj/c/common/dictionary.o bin/obj/c/dec/state.o bin/obj/c/dec/huffman.o bin/obj/c/dec/decode.o bin/obj/c/dec/bit_reader.o bin/obj/c/enc/literal_cost.o bin/obj/c/enc/compress_fragment.o bin/obj/c/enc/encoder_dict.o bin/obj/c/enc/brotli_bit_stream.o bin/obj/c/enc/dictionary_hash.o bin/obj/c/enc/utf8_util.o bin/obj/c/enc/histogram.o bin/obj/c/enc/backward_references.o bin/obj/c/enc/encode.o bin/obj/c/enc/cluster.o bin/obj/c/enc/static_dict.o bin/obj/c/enc/metablock.o bin/obj/c/enc/block_splitter.o bin/obj/c/enc/memory.o bin/obj/c/enc/compress_fragment_two_pass.o bin/obj/c/enc/backward_references_hq.o bin/obj/c/enc/entropy_encode.o bin/obj/c/enc/bit_cost.o
Step #15 - "compile-afl-address-x86_64": make[1]: Leaving directory '/src/woff2/brotli'
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -stdlib=libc++   src/font.o  src/glyph.o  src/normalize.o  src/table_tags.o  src/transform.o  src/woff2_dec.o  src/woff2_enc.o  src/woff2_common.o  src/woff2_out.o  src/variable_length.o brotli/bin/obj/c/common/*.o brotli/bin/obj/c/enc/*.o brotli/bin/obj/c/dec/*.o src/woff2_compress.o -o woff2_compress
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -stdlib=libc++   src/font.o  src/glyph.o  src/normalize.o  src/table_tags.o  src/transform.o  src/woff2_dec.o  src/woff2_enc.o  src/woff2_common.o  src/woff2_out.o  src/variable_length.o brotli/bin/obj/c/common/*.o brotli/bin/obj/c/enc/*.o brotli/bin/obj/c/dec/*.o src/woff2_decompress.o -o woff2_decompress
Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -stdlib=libc++   src/font.o  src/glyph.o  src/normalize.o  src/table_tags.o  src/transform.o  src/woff2_dec.o  src/woff2_enc.o  src/woff2_common.o  src/woff2_out.o  src/variable_length.o brotli/bin/obj/c/common/*.o brotli/bin/obj/c/enc/*.o brotli/bin/obj/c/dec/*.o src/woff2_info.o -o woff2_info
Step #15 - "compile-afl-address-x86_64": clang-14clang-14: : error: no such file or directory: 'src/woff2_dec.o'error: 
Step #15 - "compile-afl-address-x86_64": no such file or directory: 'src/woff2_dec.o'
Step #15 - "compile-afl-address-x86_64": clang-14: error: no such file or directory: 'src/woff2_dec.o'
Step #15 - "compile-afl-address-x86_64": make: *** [Makefile:56: woff2_compress] Error 1
Step #15 - "compile-afl-address-x86_64": make: *** Waiting for unfinished jobs....
Step #15 - "compile-afl-address-x86_64": make: *** [Makefile:56: woff2_decompress] Error 1
Step #15 - "compile-afl-address-x86_64": make: *** [Makefile:56: woff2_info] Error 1
Step #15 - "compile-afl-address-x86_64": **********************************************************
```
I wonder why this is?

**Comment by jonathanmetzman** [comment]
CC @vanhauser-thc fingers crossed.

**Comment by vanhauser-thc** [comment]
Haha :)

**Comment by jonathanmetzman** [comment]
> Only woff2 build fails:
> 
> ```
> Step #15 - "compile-afl-address-x86_64":         -c c/enc/block_splitter.c -o bin/obj/c/enc/block_splitter.o
> Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
> Step #15 - "compile-afl-address-x86_64":         -c c/enc/memory.c -o bin/obj/c/enc/memory.o
> Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
> Step #15 - "compile-afl-address-x86_64":         -c c/enc/compress_fragment_two_pass.c -o bin/obj/c/enc/compress_fragment_two_pass.o
> Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
> Step #15 - "compile-afl-address-x86_64":         -c c/enc/backward_references_hq.c -o bin/obj/c/enc/backward_references_hq.o
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
> Step #15 - "compile-afl-address-x86_64":         -c c/enc/entropy_encode.c -o bin/obj/c/enc/entropy_encode.o
> Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
> Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -fno-omit-frame-pointer   -D __STDC_FORMAT_MACROS -fno-tree-vrp -O2  -Ic/include \
> Step #15 - "compile-afl-address-x86_64":         -c c/enc/bit_cost.c -o bin/obj/c/enc/bit_cost.o
> Step #15 - "compile-afl-address-x86_64": clang-14: warning: optimization flag '-fno-tree-vrp' is not supported [-Wignored-optimization-argument]
> Step #15 - "compile-afl-address-x86_64": rm -f libbrotli.a
> Step #15 - "compile-afl-address-x86_64": ar -crs libbrotli.a bin/obj/c/common/transform.o bin/obj/c/common/dictionary.o bin/obj/c/dec/state.o bin/obj/c/dec/huffman.o bin/obj/c/dec/decode.o bin/obj/c/dec/bit_reader.o bin/obj/c/enc/literal_cost.o bin/obj/c/enc/compress_fragment.o bin/obj/c/enc/encoder_dict.o bin/obj/c/enc/brotli_bit_stream.o bin/obj/c/enc/dictionary_hash.o bin/obj/c/enc/utf8_util.o bin/obj/c/enc/histogram.o bin/obj/c/enc/backward_references.o bin/obj/c/enc/encode.o bin/obj/c/enc/cluster.o bin/obj/c/enc/static_dict.o bin/obj/c/enc/metablock.o bin/obj/c/enc/block_splitter.o bin/obj/c/enc/memory.o bin/obj/c/enc/compress_fragment_two_pass.o bin/obj/c/enc/backward_references_hq.o bin/obj/c/enc/entropy_encode.o bin/obj/c/enc/bit_cost.o
> Step #15 - "compile-afl-address-x86_64": make[1]: Leaving directory '/src/woff2/brotli'
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -stdlib=libc++   src/font.o  src/glyph.o  src/normalize.o  src/table_tags.o  src/transform.o  src/woff2_dec.o  src/woff2_enc.o  src/woff2_common.o  src/woff2_out.o  src/variable_length.o brotli/bin/obj/c/common/*.o brotli/bin/obj/c/enc/*.o brotli/bin/obj/c/dec/*.o src/woff2_compress.o -o woff2_compress
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -stdlib=libc++   src/font.o  src/glyph.o  src/normalize.o  src/table_tags.o  src/transform.o  src/woff2_dec.o  src/woff2_enc.o  src/woff2_common.o  src/woff2_out.o  src/variable_length.o brotli/bin/obj/c/common/*.o brotli/bin/obj/c/enc/*.o brotli/bin/obj/c/dec/*.o src/woff2_decompress.o -o woff2_decompress
> Step #15 - "compile-afl-address-x86_64": /src/aflplusplus/afl-clang-fast++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope  -stdlib=libc++   src/font.o  src/glyph.o  src/normalize.o  src/table_tags.o  src/transform.o  src/woff2_dec.o  src/woff2_enc.o  src/woff2_common.o  src/woff2_out.o  src/variable_length.o brotli/bin/obj/c/common/*.o brotli/bin/obj/c/enc/*.o brotli/bin/obj/c/dec/*.o src/woff2_info.o -o woff2_info
> Step #15 - "compile-afl-address-x86_64": clang-14clang-14: : error: no such file or directory: 'src/woff2_dec.o'error: 
> Step #15 - "compile-afl-address-x86_64": no such file or directory: 'src/woff2_dec.o'
> Step #15 - "compile-afl-address-x86_64": clang-14: error: no such file or directory: 'src/woff2_dec.o'
> Step #15 - "compile-afl-address-x86_64": make: *** [Makefile:56: woff2_compress] Error 1
> Step #15 - "compile-afl-address-x86_64": make: *** Waiting for unfinished jobs....
> Step #15 - "compile-afl-address-x86_64": make: *** [Makefile:56: woff2_decompress] Error 1
> Step #15 - "compile-afl-address-x86_64": make: *** [Makefile:56: woff2_info] Error 1
> Step #15 - "compile-afl-address-x86_64": **********************************************************
> ```
> 
> I wonder why this is?

CC @rsheeter

**Comment by vanhauser-thc** [comment]
@rsheeter @jonathanmetzman 
weird thing.
the build is doing `-j$(nproc)`, and in the parallel build the compilation of src/woff2_dec.o is not done yet when it is needed.

fix: compiles for me hard  setting -j4 (or removing -j) in build.sh. or you could tell the woff2 Makefile the necessary dependency for the pipeline.

**Pull request merged**